packages/config/src/merge.js

```diff
@@ -1,41 +1,42 @@
 'use strict'
 
-const { throwError } = require('./error')
+const { groupBy } = require('./utils/group')
 const { mergeConfigs } = require('./utils/merge')
 
 // Merge `--defaultConfig` which is used to retrieve UI build settings and
 // UI-installed plugins.
 // Also merge `inlineConfig` which is used for Netlify CLI flags.
+// Context-specific plugins can only be specified in the `config`, so it ok to
+// handle them later.
 const mergeAllConfigs = function ([defaultConfig, config, inlineConfig]) {
   const plugins = mergePlugins([defaultConfig, config, inlineConfig])
   const configA = mergeConfigs([defaultConfig, config, inlineConfig])
   return { ...configA, plugins }
 }
 
-// `defaultConfig` `plugins` are UI-installed plugins. Those are merged by
-// concatenation instead of override.
-// Also, we add the `origin` field of plugins: installed through `ui` or through
-// `config` (`netlify.toml`)
-// Context-specific plugins can only be specified in the `config`, so it ok to
-// handle them later.
+// TODO: move this logic to `deepMerge()` instead so it used during context
+// merging, allowing context-specific plugins.
 const mergePlugins = function (configs) {
-  return configs.flatMap(getPlugins).filter(isNotOverridenPlugin)
+  return configs.map(getPlugins).reduce(mergePluginsArrays, [])
 }
 
 const getPlugins = function ({ plugins = [] }) {
   return plugins
 }
 
-// When a plugin is specified both in the UI and netlify.toml, we only keep
-// the netlify.toml one.
-const isNotOverridenPlugin = function (plugin, index, plugins) {
-  const overridingPlugin = plugins.slice(index + 1).find((pluginA) => plugin.package === pluginA.package)
+// Merge two `config.plugins`. Merge plugins with the same `plugin.package`.
+const mergePluginsArrays = function (pluginsA, pluginsB) {
+  const plugins = [...pluginsA, ...pluginsB]
+  return groupBy(plugins, 'package').map(mergePluginConfigs)
```

Comment:
We don't allow `package = "plugin@v1"` and `package = "plugin@v2"` correct?
If we do, should we keep both instances (like the current code does)?

Reply:
No, `package` is only the package name. It does not include any version identifier at the moment. (I just double checked)